docs/Design/open telemetry.md

```diff
@@ -0,0 +1,218 @@
+implementing OTel into the system using SDK's directly. With manual
+instrumentation, you get full control of what is happening but it will be more
+complex and time consuming to get everything in place. OTel does provide SDK's
+for JavaScript with Typescript bindings.
+
+OTel also offers automatic instrumentation, which can hook into existing
+libraries providing basic instrumentation with just a few lines of code. To test
+this out, I added the following code to the initialize function of the
+Meadowlark-core frontendFacade
+
+```JavaScript
+  const sdk = new NodeSDK({
+    traceExporter: new ConsoleSpanExporter(),
+    metricReader: new PeriodicExportingMetricReader({
+      exporter: new ConsoleMetricExporter(),
+    }),
+    instrumentations: [getNodeAutoInstrumentations()],
+  });
+
+  sdk.start();
+```
+
+and with a call to POST a student to the Meadowlark API, we get the following
+back in the traces:
+
+```JavaScript
+{
+  traceId: 'c165a7a72532438262039f0ca57e57c4',
+  parentId: '2cb405fe86791bf0',
+  traceState: undefined,
+  name: 'PUT',
+  id: '5dabc5f64a017428',
+  kind: 2,
+  timestamp: 1683667015049000,
+  duration: 439410,
+  attributes: {
+    'http.url': 'http://localhost:9200/ed-fi%243-3-1-b%24student/_doc/bd4acc3c-af8d-4da3-88ea-e232d3068427?refresh=true',
+    'http.method': 'PUT',
+    'http.target': '/ed-fi%243-3-1-b%24student/_doc/bd4acc3c-af8d-4da3-88ea-e232d3068427?refresh=true',
+    'net.peer.name': 'localhost',
+    'http.host': 'localhost:9200',
+    'http.user_agent': 'opensearch-js/2.1.0 (win32 10.0.19045-x64; Node.js v16.14.0)',
+    'net.peer.ip': '127.0.0.1',
+    'net.peer.port': 9200,
+    'http.response_content_length_uncompressed': 229,
+    'http.status_code': 201,
+    'http.status_text': 'CREATED',
+    'http.flavor': '1.1',
+    'net.transport': 'ip_tcp'
+  },
+  status: { code: 0 },
+  events: [],
+  links: []
+}
+{
+  traceId: 'c165a7a72532438262039f0ca57e57c4',
+  parentId: undefined,
+  traceState: undefined,
+  name: 'POST',
+  id: '2cb405fe86791bf0',
+  kind: 1,
+  timestamp: 1683667014784000,
+  duration: 707563,
+  attributes: {
+    'http.url': 'http://localhost:3000/local/v3.3b/ed-fi/students/',
+    'http.host': 'localhost:3000',
+    'net.host.name': 'localhost',
+    'http.method': 'POST',
+    'http.scheme': 'http',
+    'http.target': '/local/v3.3b/ed-fi/students/',
+    'http.user_agent': 'vscode-restclient',
+    'http.request_content_length_uncompressed': 204,
+    'http.flavor': '1.1',
+    'net.transport': 'ip_tcp',
+    'net.host.ip': '127.0.0.1',
+    'net.host.port': 3000,
+    'net.peer.ip': '127.0.0.1',
+    'net.peer.port': 57641,
+    'http.status_code': 201,
+    'http.status_text': 'CREATED'
+  },
+  status: { code: 0 },
+  events: [],
+  links: []
+}
+
+```
+
+The first trace coming from the POST to the students endpoint to create a
+student. The second trace is the call adding the new student to Opensearch
```

Comment:
Interesting that they are in reverse order. Is that just because it is an asynchronous process?

Reply:
Discussed 1:1 on Friday, but unknown to the reason